curves/bn254/icicle.go

```diff
@@ -1,105 +1,135 @@
 package bn254
 
 import (
-	"fmt"
-	"math"
-	"unsafe"
-
-	"github.com/consensys/gnark-crypto/ecc/bn254"
-	"github.com/consensys/gnark-crypto/ecc/bn254/fp"
-	goicicle "github.com/ingonyama-zk/icicle/goicicle"
-	icicle "github.com/ingonyama-zk/icicle/goicicle/curves/bn254"
+	"errors"
+
+	bn254 "github.com/consensys/gnark-crypto/ecc/bn254"
+	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	"github.com/ingonyama-zk/icicle/wrappers/golang/core"
+	cr "github.com/ingonyama-zk/icicle/wrappers/golang/cuda_runtime"
+	icicle_bn254 "github.com/ingonyama-zk/icicle/wrappers/golang/curves/bn254"
 )
 
-type OnDeviceData struct {
-	P    unsafe.Pointer
-	Size int
+func MsmOnDevice(gnarkPoints []bn254.G1Affine, gnarkScalars []fr.Element) (*bn254.G1Affine, error) {
+	iciclePoints := HostSliceFromPoints(gnarkPoints)
+	icicleScalars := HostSliceFromScalars(gnarkScalars)
+
+	cfg := core.GetDefaultMSMConfig()
+	var p icicle_bn254.Projective
+	var out core.DeviceSlice
+	_, e := out.Malloc(p.Size(), p.Size())
+	if e != cr.CudaSuccess {
+		return nil, errors.New("cannot allocate")
+	}
+	e = icicle_bn254.Msm(icicleScalars, iciclePoints, &cfg, out)
+	if e != cr.CudaSuccess {
+		return nil, errors.New("msm failed")
+	}
+	outHost := make(core.HostSlice[icicle_bn254.Projective], 1)
+	outHost.CopyFromDevice(&out)
+	out.Free()
+	return ProjectiveToGnarkAffine(&outHost[0]), nil
 }
 
-func INttOnDevice(scalars_d, twiddles_d, cosetPowers_d unsafe.Pointer, size, sizeBytes int, isCoset bool) unsafe.Pointer {
-	ReverseScalars(scalars_d, size)
+func Ntt[T any](gnarkScalars fr.Vector, dir core.NTTDir, cfg *core.NTTConfig[T]) (fr.Vector, error) {
+	icicleScalars := core.HostSliceFromElements(BatchConvertFromFrGnark(gnarkScalars))
```

Comment:
For ntt, we really only need to convert from uint64 to uint32 since Montgomery form is supported